dev-tools/smoke.js

```diff
@@ -0,0 +1,452 @@
+			items: new Pool(onlyValid(dex.gen, dex.data.Items, i => dex.getItem(i)), this.prng),
+			abilities: new Pool(onlyValid(dex.gen, dex.data.Abilities, a => dex.getAbility(a)), this.prng),
+			moves: new Pool(onlyValid(dex.gen, dex.data.Movedex, m => dex.getMove(m),
+				m => (m !== 'struggle' && (m === 'hiddenpower' || m.substr(0, 11) !== 'hiddenpower'))), this.prng),
+		};
+	}
+
+	logProgress(p) {
+		// `\r` = return to the beginning of the line
+		// `\x1b[k` (`\e[K`) = clear all characters from cursor position to EOL
+		if (this.games) process.stdout.write('\r\x1b[K');
+		// Deliberately don't print a `\n` character so that we can overwrite
+		process.stdout.write(
+			`[${this.format}] P:${p.pokemon} I:${p.items} A:${p.abilities} M:${p.moves} = ${this.games}`);
+	}
+}
+
+module.exports = {TeamGenerator, Pool, CoordinatedPlayerAI, SmokeRunner};
+
+function onlyValid(gen, obj, getter, additional, nonStandard) {
+	return Object.keys(obj).filter(k => {
+		const v = getter(k);
+		return v.gen <= gen &&
+			(!v.isNonstandard || !!nonStandard) &&
+			(!additional || additional(k, v));
+	});
+}
+
+function getSignatures(dex, pools) {
+	const signatures = new Map();
+	for (const id of pools.items.possible) {
+		const item = dex.data.Items[id];
+		if (item.megaEvolves) {
+			const pokemon = toId(item.megaEvolves);
+			const combo = {item: id};
+			let combos = signatures.get(pokemon);
+			if (!combos) {
+				combos = [];
+				signatures.set(pokemon, combos);
+			}
+			combos.push(combo);
+		} else if (item.zMoveUser) {
+			for (const user of item.zMoveUser) {
+				const pokemon = toId(user);
+				const combo = {item: id};
+				if (item.zMoveFrom) combo.move = toId(item.zMoveFrom);
+				let combos = signatures.get(pokemon);
+				if (!combos) {
+					combos = [];
+					signatures.set(pokemon, combos);
+				}
+				combos.push(combo);
+			}
+		}
+	}
+	return signatures;
+}
+
+if (require.main === module) {
+	Dex.includeModData();
+	RejectionTracker.register();
+
+	// TODO: Add triple battles once supported by the AI.
+	const FORMATS = [
+		'gen7customgame', 'gen7doublescustomgame',
+		'gen6customgame', 'gen6doublescustomgame',
+		'gen5customgame', 'gen5doublescustomgame',
+		'gen4customgame', 'gen4doublescustomgame',
+		'gen3customgame', 'gen3doublescustomgame',
+		'gen2customgame',
+		'gen1customgame'];
+
+	// Because of our handwavy accounting of 'usage', as well as to account for ordering effects and
+	// the interplay of various combinations, we should actually run multiple cycles and exhaust our
+	// pools multiple times to provide more confidence that we're actually smoking out issues.
+	//
+	// If cycles is negative the smoke test will instead run indefinitely in case you wish to
+	// dedicate some computer resources to finding PS! edge case crashes instead of mining crypto,
+	// folding proteins, or searching for extraterrestrials/large primes.
+	const cycles = Number(process.argv[2]) || DEFAULT_CYCLES;
+	const forever = cycles < 0;
```

Comment:
It would be nice to then make `cycles` positive here, instead of needing that `Math.abs` in a weird spot.

Reply:
Yeah, but I prefer the code this way so I can leave `cycles` `const` as well as be less verbose. :)

Let's argue about this some more instead of merging!

Reply:
See, this is why I don't like `const`, it forces you to write worse code because you start optimizing for this sort of thing. :p

Reply:
```js
	const cyclesSetting = Number(process.argv[2]) || DEFAULT_CYCLES;
	const forever = cyclesSetting < 0;
	const cycles = Math.abs(cyclesSetting);
```

Reply:
I mean, obviously I thought of that, I just didn't like it as because `cyclesSetting` seems contrived. You could also pass `cycle: Math.abs(cycle)` to the `SmokeRunner`, but that also kind of makes the code uglier as well IMO because the shortcut notation looks better

I don't know why I'm being stubborn about this because its so trivial, but I like it the way it is. *shrugs*. I wonder if this is just a ['just remove the  duck'](https://rachelbythebay.com/w/2013/06/05/duck/) scenario given how I'm sure the `Pool` logic etc is more complicated and more worth this sort of attention.

Reply:
You wouldn't be resisting this much if this were actually a duck. I just don't like that `cycles` isn't actually a cycle count, only its absolute value is. Hence why I think `cyclesSetting` is a better name for it.

Reply:
I guess it's not a huge deal to me in that like most things, if you insisted, I'd let it go. I'm just not seeing the value here. It seems like the variable names are bad for seemingly no reason.

Reply:
In my experience, good variable names tends to be one of the more important things for readability.

Reply:
Only now that I've removed the `Math.abs` did I remember why I needed it - the error message on line 340 doesn't match the command run. Doesnt really matter, because you should be able to use negative or positive to repro, but now I at least remember *why* I wrote it the way I did.